### DISCUSSION TITLE
Change where custom fields are stored

### DISCUSSION THREAD

**Discussion by brianteeman** [joomla/joomla-cms#38658] [2. Cold Features]
## Today
The current database structure for custom fields is
- The structure of the custom field is stored in #__fields
- The content of the field is stored in #__field_values

#__field_values has three fields
field_id ->what type of field is it links to #__fields
item_id->which article is this for
value -> content of the field

This has several well known limitations
- no versioning for the fields
- no search (might be solved now)
- no inclusion of the field in the list views
- complex searches to build the complete content item

## Tomorrow
This idea is just an idea. I don't know if its feasible but it sounds like it should be.

When you create a field is would still save the field structure in #__fields
**BUT and this is the critical part**
When you create the field it also adds a field to the #__content table

So now the field data is stored inside the content item. This would immediately resolve the 4 limitations above.

## Backwards compatibility
Probably zero 
but you could easily have a script that would add the fields to #__content and then move the data from #__field_values

**Comment by brianteeman** [comment]
An alternative would be to add an additional field to the #__content table at install/upgrade and to store all the field data as json in that field

**Comment by denysdesign** [comment]
Nice idea

**Comment by b2z** [comment]
Historically the content table is designed for articles. Now you want to place there the content from all other entities including extensions that support custom fields. Does not sound so good.

**Reply by brianteeman** [reply]
Where did I say that. It would be a stupid idea to store contact fields in a content table

**Reply by b2z** [reply]
> When you create the field it also adds a field to the #__content table

Anyway what 3rd party extensions should do?

**Reply by brianteeman** [reply]
exactly the same create the extra field in their own component table

**Reply by b2z** [reply]
So it is a b/c then.

**Reply by brianteeman** [reply]
was I not clear
![image](https://user-images.githubusercontent.com/1296369/187862097-953dc0cd-c7aa-4f20-8195-1db4ec4876d1.png)

**Reply by b2z** [reply]
Not really. Still require effort from 3rd party extensions, because in one moment fields will be broken.

**Reply by brianteeman** [reply]
I honestly dont know how I could have made it more clear

**Reply by brianteeman** [reply]
> Not really. Still require effort from 3rd party extensions, because in one moment fields will be broken.

only if it was forced on everyone without notice or options to optin etc etc

please dont treat me like an idiot

**Comment by sergeytolkachyov** [comment]
You need to remember that Joomla is updated separately, and third - party extensions are updated separately. If the change happens in Joomla, and the third - party extension developers do not make changes to their extensions **in advance**, then there will be a lot of negativity. 
There are things that are done very easily if done in a timely manner. And people who know what they're doing. Often sites are administered by people who can only click with the mouse and nothing more. To make changes that break backward compatibility, you need **to advance for a long time (several months)**  and notify developers and site administrators more than once. To do this, use system messages in the admin panel, notifications to the site administrator's email. Then there is a chance that this really important notification will attract attention among the tons of information pouring out on us every minute. 
And those who do not know how to work with the code will find those who can **in advance**. This will allow you to take care of users, because this is the ultimate goal of any business.

**Reply by brianteeman** [reply]
Please dont treat me like an idiot!!

**Reply by sergeytolkachyov** [reply]
Nobody takes anyone for an idiot. Idiots can't work with Joomla. And even quite mediocre people can't work, because you need to think and think well with her. Someone has 2 fields on the site for the entire project, and someone may have several thousand fields (catalogs of cars, car parts, apartments, etc.). You need to assume a variety of implementation scenarios and how this could potentially affect the final owners of the site.

**Reply by brianteeman** [reply]
of course. 

maybe I should have assumed the people reading this were not very clever and written a ten thousand word proposal to explain a simple idea and address every little thing at once. but then no one would have read it. 

dont teach me to suck eggs

**Comment by Fedik** [comment]
Sorry Brian, this is a bad idea, it will give a lot of headaches for maintain CMS, I mean maintaining DB structure. This is the reason why other CMS use dedicated table for custom fields.

**Reply by brianteeman** [reply]
even the alternative option where the field always exists and the data is stored as json?

**Reply by Fedik** [reply]
json in general hard to use, for filter query. 
However, having some extra column for History table and dump all extra fields as json there, can be not that bad idea.

**Reply by brianteeman** [reply]
looking at drupal they have a table for each field and a second table to the revisions
creating a new field creates a new pair of tables

**Reply by Fedik** [reply]
Yeah, they have a very complex structure :)

**Reply by Fedik** [reply]
There, if you have 100 extra fields there, you will have 100 extra SQL JOIN, to load them all.

**Reply by Fedik** [reply]
Hm, or we probably can do `#__field_values_history`, kind of. 
Not a simple task, but shold be possible also.

**Reply by brianteeman** [reply]
exactly what drupal does

**Comment by Septdir** [comment]
It would be a great idea to get rid of EAV, and it would give a huge performance boost, but this is only possible if the field values are stored in JSON. And in order to work normally with JSON, you need to increase the requirements for databases to Mysql 8.0.21

Well, as for compatibility with the old structure, this is solved with the help of legacy and the update script. But no one listens to me anyway, so this is just my opinion.

**Reply by richard67** [reply]
> And in order to work normally with JSON, you need to increase the requirements for databases to Mysql 8.0.21

What do you mean with that? The JSON data type? If so: Joomla cannot use that because Joomla has to work with MariaDB, too, and MariaDB does not have the JSON data type.

**Reply by richard67** [reply]
See https://mariadb.com/kb/en/json-data-type/ for the details.

**Reply by Septdir** [reply]
> > And in order to work normally with JSON, you need to increase the requirements for databases to Mysql 8.0.21
> 
> What do you mean with that? The JSON data type? If so: Joomla cannot use that because Joomla has to work with MariaDB, too, and MariaDB does not have the JSON data type.

Maria works great with json and has all the features you need.  I use MariaDB myself, Maria 10.4.1 will be required to work correctly

**Reply by progreccor** [reply]
> See https://mariadb.com/kb/en/json-data-type/ for the details.

actual version of MariaDB is 10.6 and it works with JSON like a charm

**Comment by alikon** [comment]
joomla already manage JSON _"in some way"_ from the core
 https://github.com/joomla/joomla-cms/blob/4.2-dev/installation/sql/mysql/base.sql#L16
as a silly example

so imho it's not matter of **WHERE**

much more of "**WHAT**" we need to store/retriieve

**Comment by jeckodevelopment** [comment]
Using your reasoning, we should store field content to "destination tables", so if it's a field linked to Category, you have the field content in the categories table, if you have a field for contacts, you have the content stored in the contact table.
Wouldn't be better to solve the issues that you pointed out keeping the current database structure? (i'm also wondering the impact on third-party extensions that implemented fields).
